feat:[CDS-88594]: Trigger maps parameters to Pipeline Runtime Inputs and enable reuse of InputSets (#4969)
harness/developer-hub: docs/platform/git-experience/manage-input-sets-in-simplified-git-experience.md

For more details, go to [Manage a Harness Pipeline Repo Using Git Experience](manage-a-harness-pipeline-repo-using-git-experience.md).

> ### Customizing Input Configuration in Trigger YAML through Override YAML

Reviewer: ```suggestion
### Customize trigger input configuration using override YAML
```

> Suppose you have several input sets that you wish to employ both manually and via triggers. However, you desire the capability to specify a different value for the trigger, allowing you to override a particular parameter when utilizing triggers.

Reviewer: ```suggestion
Suppose you have several input sets that you want to employ both manually and via triggers. However, you want the ability to specify a different value for the trigger, allowing you to override a particular parameter when using triggers.
```

> For example: if a user has a input set ``inputSet1`` like this:

Reviewer: ```suggestion
For example, if you have an input set `inputSet1` like the following:
```

> ```

Reviewer: ```suggestion
```yaml
```

inputSet:
  name: i1
  tags: {}
  identifier: inputSet1
  orgIdentifier: default
  projectIdentifier: default_project
  pipeline:
    identifier: trigger_overrides
    variables:
      - name: TAG
        type: String
        value: test_trigger_override
```

> For triggers, users can effortlessly customize the value of the TAG by extending the Trigger YAML. Following the ``inputSetRefs`` section, users can now include an ``overrideInputs`` section, allowing them to specify and update the value of the variable they intend to override.

Reviewer: ```suggestion
For triggers, you can customize the value of the TAG by extending the trigger YAML. Following the `inputSetRefs` section, you can include an `overrideInputs` section that allows you to specify and update the value of the variable you want to override.
```

> Reference Yaml:

Reviewer: ```suggestion
#### Reference YAML
```

> ```

Reviewer: ```suggestion
```yaml
```

trigger:
  name: test1
  identifier: test1
  enabled: true
  description: ""
  tags: {}
  stagesToExecute: []
  orgIdentifier: default
  projectIdentifier: default_project
  pipelineIdentifier: trigger_overrides
  source:
    type: Webhook
    spec:
      type: Custom
      spec:
        payloadConditions: []
        headerConditions: []
  inputSetRefs:
    - i1
  overrideInputs: |
    pipeline:
      identifier: trigger_overrides
      variables:
        - name: TAG
          type: String
          value: test_trigger_override
```

:::info note

> When overrideInputs are provided, the user-supplied values take precedence over all other configurations. In other words, any values specified in the overrideInputs section will override and take preference over the corresponding values from other configurations.

Reviewer: ```suggestion
When `overrideInputs` are supplied, their values take precedence over other configurations. This means that values specified in the `overrideInputs` section will override the corresponding values from other configurations.
```

For more details, go to [Manage a Harness Pipeline Repo Using Git Experience](manage-a-harness-pipeline-repo-using-git-experience.md).

> ### Customize trigger input configuration using override YAML

Reviewer 2: Why is this information inside the "Simplified Git Experience" topic? Shouldn't it be a standalone topic under the Triggers section at https://developer.harness.io/docs/category/triggers ? 
Please create a new topic there to explain this flow in detail.

> Suppose you have several input sets that you wish to use both manually and via triggers. However, you desire the capability to specify a different value for the trigger, allowing you to override a particular parameter when utilizing triggers.

Reviewer 2: We don't want to specify "a different value for the trigger". We want to override **some** of the values in the Input Set when that Input Set is being used in the Trigger. So, if the Input Set includes the Service Name as a value, we want to not use the Service Name value that's in the Input Set, and instead use some Service Name that is provided through the Trigger's Payload (or else another hardcoded Service Name).

> Suppose you have several input sets that you wish to use both manually and via triggers. However, you desire the capability to specify a different value for the trigger, allowing you to override a particular parameter when utilizing triggers.

Reviewer 2: Can you expand on the use case a bit? When does the user need to do any of this?

> You can do the following so as to define overriden value in the Trigger Yaml:

Reviewer 2: "YAML" is always written in all caps, please.

> 1. You need to define a section``inputYaml`` in trigger YAML as depicted in the sample below. 

Reviewer 2: Why can't it be done in the UI? If this is only supported in the YAML we need to call it out.

> 2. Please note that it is important to structure the ``inputYaml`` similar to what is defined in the pipeline and it is important to note that you provide correct ``identifier``. 

Reviewer 2: When you say "similar to what is defined in the pipeline", where is the Input YAML in the pipeline that the user needs to copy from?

> 2. Please note that it is important to structure the ``inputYaml`` similar to what is defined in the pipeline and it is important to note that you provide correct ``identifier``. 

Reviewer 2: "identifier" of what?

> Let's have a look at a sample yaml:-

Reviewer 2: Is this a sample YAML of Trigger? Pipeline? Input Set? It's unclear. We should be providing all of these for the user to get the full context.

  pipelineIdentifier: sample_svc_cd
  source:
    type: Webhook
    spec:
      type: Custom
      spec:
        payloadConditions: []
        headerConditions: []
  pipelineBranchName: master
  inputYaml: |
    pipeline:
      template:
        templateInputs:
          stages:
            - stage:
                identifier: deployment_setup
                template:
                  templateInputs:
                    type: Deployment
                    spec:
                      service:
                        serviceRef: dummy_svc
                        serviceInputs:
                          serviceDefinition:
                            type: Kubernetes
                            spec:
                              artifacts:
                                primary:
                                  primaryArtifactRef: harness_hello_world
                                  sources:
                                    - identifier: harness_hello_world
                                      type: Ecr
                                      spec:
                                        tag: <+trigger.payload.tag>
  inputSetRefs:
    - dummy_svc

```
:::info note

> 1. When ``inputYaml`` is provided, the user-supplied values take precedence over all other configurations. In other words, any values specified in the  ``inputYaml`` section will override and take preference over the corresponding values from other configurations.

Reviewer 2: What "other configurations" are there? Did you mean other Input Sets?

> 2. The order of input values doesn't matter, but as mentioned above the structure of the ``inputYaml`` is very important.

Reviewer 2: We need some more examples of how to create the Input YAML - where to look, how it corresponds to what is to be overriden, how to select which part of the Input YAML to take,...

> :::

Reviewer 2: Please add a section on common error messages, debugging strategies for those cases.
